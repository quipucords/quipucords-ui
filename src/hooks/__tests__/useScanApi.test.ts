import { renderHook } from '@testing-library/react';
import axios from 'axios';
import {
  useCreateScanApi,
  useDeleteScanApi,
  useDownloadReportApi,
  useGetScanJobsApi,
  useRunScanApi,
  useShowConnectionsApi
} from '../useScanApi';

describe('useShowConnectionsApi', () => {
  let hookResult;

  beforeEach(() => {
    const hook = renderHook(() => useShowConnectionsApi());
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an api call to show connections', () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'get');

    apiCall({ id: 123, connection: { id: 456 } });
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while fetching connections', async () => {
    const { showConnections } = hookResult;
    jest.spyOn(axios, 'get').mockResolvedValue({});

    const response = await showConnections({ id: 123, connection: { id: 456 } });
    expect(response).toMatchSnapshot('showConnections success');
  });

  it('should handle errors while fetching connections', async () => {
    const { showConnections } = hookResult;
    jest.spyOn(axios, 'get').mockRejectedValue({ isAxiosError: true, message: 'Mock error' });

    await expect(showConnections({ id: 123, connection: { id: 456 } })).rejects.toMatchSnapshot(
      'showConnections error'
    );
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    expect(
      callbackSuccess({
        data: {
          results: [
            { status: 'success', name: 'Connection 1' },
            { status: 'failed', name: 'Connection 2' },
            { status: 'unreachable', name: 'Connection 3' }
          ]
        }
      })
    ).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', async () => {
    const { callbackError } = hookResult;

    await expect(
      callbackError({
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      })
    ).rejects.toMatchSnapshot('callbackError');
  });
});

describe('useDeleteScanApi', () => {
  let mockOnAddAlert;
  let hookResult;

  beforeEach(() => {
    mockOnAddAlert = jest.fn();
    const hook = renderHook(() => useDeleteScanApi(mockOnAddAlert));
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an api call to delete scans', () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'post');

    apiCall([456, 789]).catch(Function.prototype);
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to delete single and multiple scans', async () => {
    const { deleteScans } = hookResult;
    jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve({}));

    await deleteScans({ id: 123, name: 'lorem ipsum dolor sit' });
    await deleteScans([
      { id: 456, name: 'lorem ipsum' },
      { id: 789, name: 'dolor sit' }
    ]);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('deleteScans, success');
  });

  it('should handle errors while attempting to delete single and multiple scans', async () => {
    const { deleteScans } = hookResult;
    jest.spyOn(axios, 'post').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await deleteScans({ id: 123, name: 'lorem ipsum dolor sit' });
    await deleteScans([
      { id: 456, name: 'lorem ipsum' },
      { id: 789, name: 'dolor sit' }
    ]);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('deleteScans, error');
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    callbackSuccess(
      {
        data: {
          deleted: [123, 456]
        }
      },
      [
        { id: 123, name: 'Lorem' },
        { id: 456, name: 'Ipsum' }
      ]
    );

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', () => {
    const { callbackError } = hookResult;

    callbackError(
      {
        response: {
          data: {
            detail: 'Lorem ipsum'
          }
        }
      },
      [
        { id: 123, name: 'Lorem' },
        { id: 456, name: 'Ipsum' }
      ]
    );

    callbackError(
      {
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      },
      [
        { id: 123, name: 'Lorem' },
        { id: 456, name: 'Ipsum' }
      ]
    );

    callbackError(
      {
        message: 'Amet'
      },
      [
        { id: 123, name: 'Lorem' },
        { id: 456, name: 'Ipsum' }
      ]
    );

    callbackError({}, [
      { id: 123, name: 'Lorem' },
      { id: 456, name: 'Ipsum' }
    ]);

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackError');
  });
});

describe('useCreateScanApi', () => {
  let mockOnAddAlert;
  let hookResult;

  beforeEach(() => {
    mockOnAddAlert = jest.fn();
    const hook = renderHook(() => useCreateScanApi(mockOnAddAlert));
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an api call to create a scan', () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'post');

    apiCall({ name: 'Lorem' });
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to create a scan', async () => {
    const { createScans } = hookResult;
    jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve({}));

    await createScans({ name: 'Lorem' });
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('createScans, success');
  });

  it('should handle errors while attempting to create a scan', async () => {
    const { createScans } = hookResult;
    jest.spyOn(axios, 'post').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await createScans({ name: 'Lorem' });
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('createScans, error');
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    callbackSuccess({
      data: { name: 'Lorem' }
    });

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', () => {
    const { callbackError } = hookResult;

    callbackError(
      {
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      },
      'Lorem Ipsum'
    );

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackError');
  });
});

describe('useGetScanJobsApi', () => {
  let mockOnAddAlert;
  let hookResult;

  beforeEach(() => {
    mockOnAddAlert = jest.fn();
    const hook = renderHook(() => useGetScanJobsApi(mockOnAddAlert));
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an api call to retrieve a scanJob', () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'get');

    apiCall(1);
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to retrieve a scanJob', async () => {
    const { getScanJobs } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));

    await getScanJobs(1);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('getScanJobs, success');
  });

  it('should handle errors while attempting to retrieve a scanJob', async () => {
    const { getScanJobs } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await getScanJobs(1);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('getScanJobs, error');
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    expect(
      callbackSuccess({
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      })
    ).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', () => {
    const { callbackError } = hookResult;

    callbackError(
      {
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      },
      1
    );

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackError');
  });
});

describe('useRunScanApi', () => {
  let mockOnAddAlert;
  let hookResult;
  let mockCreateScan;

  beforeEach(() => {
    mockOnAddAlert = jest.fn();
    mockCreateScan = jest.fn().mockResolvedValue({ id: '123' });
    const mockUseCreateScanApi = jest.fn().mockReturnValue({
      createScans: mockCreateScan
    });
    const hook = renderHook(() => useRunScanApi(mockOnAddAlert, mockUseCreateScanApi));
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an API call to run a scan', async () => {
    const { apiCall } = hookResult;
    const spyAxiosRun = jest.spyOn(axios, 'post');

    await apiCall(123);
    expect(spyAxiosRun.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to run a scan', async () => {
    const { runScans } = hookResult;
    jest.spyOn(axios, 'post').mockResolvedValue(Promise.resolve({}));

    await runScans({ name: 'Lorem' });
    await runScans({ id: '123', name: 'Lorem' }, true);

    expect(mockCreateScan).toHaveBeenCalledTimes(1);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('runScans, success');
  });

  it('should handle errors while attempting to run a scan', async () => {
    const { runScans } = hookResult;
    jest.spyOn(axios, 'post').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await runScans({ name: 'Lorem' });
    await runScans({ id: '123', name: 'Lorem' }, true);

    expect(mockCreateScan).toHaveBeenCalledTimes(1);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('runScans, error');
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    callbackSuccess('Lorem');

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', () => {
    const { callbackError } = hookResult;

    callbackError(
      {
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      },
      'Lorem'
    );

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackError');
  });

  it('should handle missing scan ID and trigger custom error', async () => {
    const { runScans } = hookResult;
    mockCreateScan.mockResolvedValue({});

    await runScans({ name: 'Lorem' });
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('runScans, no ID returned');
  });
});

describe('useDownloadReportApi', () => {
  let mockOnAddAlert;
  let hookResult;

  beforeEach(() => {
    mockOnAddAlert = jest.fn();
    const hook = renderHook(() => useDownloadReportApi(mockOnAddAlert));
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an API call to download a report', async () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'get');

    await apiCall(123);
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to download a report', async () => {
    const { downloadReport } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));

    await downloadReport(123);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('downloadReport, success');
  });

  it('should handle errors while attempting to download a report', async () => {
    const { downloadReport } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await downloadReport(123);
    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('downloadReport, error');
  });

  it('should process an API success response', () => {
    const { callbackSuccess } = hookResult;

    callbackSuccess(new Blob(['mock report data'], { type: 'application/gzip' }), 123);

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackSuccess');
  });

  it('should process an API error response', () => {
    const { callbackError } = hookResult;

    callbackError(
      {
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      },
      123
    );

    expect(mockOnAddAlert.mock.calls).toMatchSnapshot('callbackError');
  });
});
