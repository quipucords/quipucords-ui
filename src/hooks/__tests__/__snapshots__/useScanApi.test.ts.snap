// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCreateScanApi should attempt an api call to create a scan: apiCall 1`] = `
[
  [
    "/api/v1/scans/",
    {
      "name": "Lorem",
    },
  ],
]
`;

exports[`useCreateScanApi should handle errors while attempting to create a scan: createScans, error 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_created_error",
    "name": "Lorem",
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useCreateScanApi should handle success while attempting to create a scan: createScans, success 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_created"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useCreateScanApi should process an API error response: callbackError 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_created_error",
    "name": "Lorem Ipsum",
    "message": "Dolor sit"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useCreateScanApi should process an API success response: callbackSuccess 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_created",
    "name": "Lorem"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useDeleteScanApi should attempt an api call to delete scans: apiCall 1`] = `
[
  [
    "/api/v1/scans/bulk_delete/",
    {
      "ids": [
        456,
        789,
      ],
    },
  ],
]
`;

exports[`useDeleteScanApi should handle errors while attempting to delete single and multiple scans: deleteScans, error 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 1,
    "name": "lorem ipsum dolor sit",
    "message": "Mock error"
  }
])",
      "variant": "danger",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 2,
    "name": "lorem ipsum, dolor sit",
    "message": "Mock error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useDeleteScanApi should handle success while attempting to delete single and multiple scans: deleteScans, success 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan",
    "count": 1,
    "name": "lorem ipsum dolor sit"
  }
])",
      "variant": "success",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan",
    "count": 2,
    "name": "lorem ipsum, dolor sit"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useDeleteScanApi should process an API error response: callbackError 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 2,
    "name": "Lorem, Ipsum",
    "message": "Lorem ipsum"
  }
])",
      "variant": "danger",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 2,
    "name": "Lorem, Ipsum",
    "message": "Dolor sit"
  }
])",
      "variant": "danger",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 2,
    "name": "Lorem, Ipsum",
    "message": "Amet"
  }
])",
      "variant": "danger",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan_error",
    "count": 2,
    "name": "Lorem, Ipsum",
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useDeleteScanApi should process an API success response: callbackSuccess 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "deleted-scan",
    "count": 2,
    "name": "Lorem, Ipsum"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useDownloadReportApi should attempt an API call to download a report: apiCall 1`] = `
[
  [
    "/api/v1/reports/123/",
    {
      "responseType": "blob",
    },
  ],
]
`;

exports[`useDownloadReportApi should handle errors while attempting to download a report: downloadReport, error 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "report_download_error",
    "name": 123,
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useDownloadReportApi should handle success while attempting to download a report: downloadReport, success 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "report_downloaded",
    "name": 123
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useDownloadReportApi should process an API error response: callbackError 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "report_download_error",
    "name": 123,
    "message": "Dolor sit"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useDownloadReportApi should process an API success response: callbackSuccess 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "report_downloaded",
    "name": 123
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useGetScanJobsApi should attempt an api call to retrieve a scanJob: apiCall 1`] = `
[
  [
    "/api/v1/scans/1/jobs/",
  ],
]
`;

exports[`useGetScanJobsApi should handle errors while attempting to retrieve a scanJob: getScanJobs, error 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-jobs_fetch_error",
    "id": 1,
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useGetScanJobsApi should handle success while attempting to retrieve a scanJob: getScanJobs, success 1`] = `[]`;

exports[`useGetScanJobsApi should process an API error response: callbackError 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-jobs_fetch_error",
    "id": 1,
    "message": "Dolor sit"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useGetScanJobsApi should process an API success response: callbackSuccess 1`] = `
{
  "response": {
    "data": {
      "message": "Dolor sit",
    },
  },
}
`;

exports[`useRunScanApi should attempt an API call to run a scan: apiCall 1`] = `
[
  [
    "/api/v1/scans/123/jobs/",
  ],
]
`;

exports[`useRunScanApi should handle errors while attempting to run a scan: runScans, error 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play_error",
    "name": "Lorem",
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play_error",
    "name": "Lorem",
    "message": "Unknown error"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useRunScanApi should handle missing scan ID and trigger custom error: runScans, no ID returned 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play_error",
    "name": "Lorem",
    "message": "No ID returned\\nThe scan or rescan operation did not return an ID."
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useRunScanApi should handle success while attempting to run a scan: runScans, success 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play",
    "name": "Lorem"
  }
])",
      "variant": "success",
    },
  ],
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play",
    "name": "Lorem"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useRunScanApi should process an API error response: callbackError 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play_error",
    "name": "Lorem",
    "message": "Dolor sit"
  }
])",
      "variant": "danger",
    },
  ],
]
`;

exports[`useRunScanApi should process an API success response: callbackSuccess 1`] = `
[
  [
    {
      "title": "t([
  "toast-notifications.description",
  {
    "context": "scan-report_play",
    "name": "Lorem"
  }
])",
      "variant": "success",
    },
  ],
]
`;

exports[`useShowConnectionsApi should attempt an api call to show connections: apiCall 1`] = `
[
  [
    "/api/v1/jobs/456/connection/",
    {
      "params": {
        "ordering": "name",
        "page": 1,
        "page_size": 1000,
        "source_type": 123,
      },
    },
  ],
]
`;

exports[`useShowConnectionsApi should handle errors while fetching connections: showConnections error 1`] = `
{
  "isAxiosError": true,
  "message": "Mock error",
}
`;

exports[`useShowConnectionsApi should handle success while fetching connections: showConnections success 1`] = `
{
  "failed": [],
  "successful": [],
  "unreachable": [],
}
`;

exports[`useShowConnectionsApi should process an API error response: callbackError 1`] = `
{
  "response": {
    "data": {
      "message": "Dolor sit",
    },
  },
}
`;

exports[`useShowConnectionsApi should process an API success response: callbackSuccess 1`] = `
{
  "failed": [
    {
      "name": "Connection 2",
      "status": "failed",
    },
  ],
  "successful": [
    {
      "name": "Connection 1",
      "status": "success",
    },
  ],
  "unreachable": [
    {
      "name": "Connection 3",
      "status": "unreachable",
    },
  ],
}
`;
